### DISCUSSION TITLE
Validate a string that must be a UUIDv4?

### DISCUSSION THREAD

**Discussion by danielniccoli** [pydantic/pydantic#7023] [Question] [answered]
I have a model with a UUID4-validated field. The field must accept UUIDs (v4) as a `uuid.UUID` object, but also as a string. Simply setting the type to `UUID4` works for the objects, but invalidates any string, even if it would be a correct UUID. I replaced the type with `my_uuid: UUID4 | Annotated[str, UUID4]` in the hopes that it now accepts both, but I may not correctly understand the syntax to achieve this, because my solution validates all strings, including `"this must fail"`.

How must I change my model in order to accept both, valid `UUID4` objects and strings representing valid UUID4?

```py
import uuid
from typing import Annotated

from pydantic import UUID4, BaseModel, BeforeValidator, field_validator
from pydantic.config import ConfigDict

config = ConfigDict(strict=True)


class MyModel(BaseModel):
    model_config = config
    # my_uuid: UUID4  # Does not work on strings
    my_uuid: UUID4 | Annotated[str, UUID4]


MyModel(my_uuid=uuid.uuid4())
MyModel(my_uuid=uuid.UUID(str(uuid.uuid4())))
MyModel(my_uuid="a6433fbe-f32e-4c76-9e89-e5d0b9ecf922")
MyModel(my_uuid="a6433fbef32e4c769e89e5d0b9ecf922")

MyModel(my_uuid="this must fail")
```

**Comment by orfisko** [comment]


**Reply by danielniccoli** [reply]
Then I suggest you fix your environment ;) The code works just fine.

**Reply by PrettyWood** [reply, marked as answer]
just use `AfterValidator`
```py
my_uuid: UUID4 | Annotated[str, AfterValidator(lambda x: uuid.UUID(x, version=4))]
```
Or you can change the strictness to false and lean on coercion
```py
my_uuid: Annotated[UUID4, Strict(False)]
```

**Reply by danielniccoli** [reply]
Both very nice solutions. Thank you!
